docs/spark/how-to-guides/dotnet-interactive-udf-issue.md

```diff
@@ -0,0 +1,65 @@
+---
+title: Using UDFs in .NET for Apache Spark interactive shells.
+description: Learn how to write and call UDFs in .NET for Apache Spark interactive shells.
+ms.date: 09/24/2020
+ms.topic: conceptual
+ms.custom: mvc,how-to
+---
+
+# How to write UDFs in .NET for Apache Spark interactive shells
+
+In this article you will learn how to use UDFs in a .NET for Apache Spark interactive environment, a couple of important caveats to keep in mind, and why Broadcast Variables don't work with .NET for Apache Spark interactive experiences.
+
+## Prerequisites
+
+1. Install [.NET Interactive](https://github.com/dotnet/interactive)
+2. Install [Jupyter lab](https://jupyter.org/)
+
+## .NET for Apach Spark Interactive experience
+
+[.NET for Apache Spark](https://github.com/dotnet/spark) uses [.NET Interactive](https://devblogs.microsoft.com/dotnet/net-interactive-is-here-net-notebooks-preview-2/) to provide .NET support for interactive experience inside Spark. Kindly go through the [official docs](https://github.com/dotnet/interactive) to understand how to setup the environment to try .NET Interactive with Jupyter notebooks.
+
+Also, it is recommended to go through [this article about UDF serialization in .NET for Apache Spark](udf-guide.md) to understand what UDFs are and how they are serialized in .NET for Apache Spark.
+We will be using Jupyter Notebooks to illustrate how to use UDFs in an interactive experience.
+
+## UDF serialization in .NET Interactive
+
+In the interactive experience, a cell can be thought of as the code snippet being submitted as part of one iteration of the [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop). For example, let's take a look at the following notebook to understand what that means:
+
+![Jupyter Notebook cells](./media/dotnet-interactive/dotnet-interactive-cells.png)
+
+Each of those blocks marked by the red arrow is one cell, or one submission of code to Spark. .NET Interactive wraps each of these cells with a wrapper class of the cell submission number, to uniquely identify each cell being submitted. Now as explained in detail in [this guide](udf-guide.md), when a UDF that references a custom object is being serialized its target is also picked up for serialization, which in the case of .NET Interactive gets wrapped by the wrapper class of the cell where the custom object is defined.
+Now let's see how that affects our UDF definition in the notebook:
+
+![UDF serialization error](./media/dotnet-interactive/udf-serialization-error.png)
+
+As can be seen in the case of `udf2_fails`, we see the error message which says Type `Submission#7` is not marked as serializable, this is because how .NET Interactive works is it wraps every object defined in a cell with its `Submission#` class which is generated on the fly and hence is not marked as `Serializable`, hence the error.
```

Comment:
See the second inline snippet `Submission#`. Is that right?

Reply:
Yes, it refers to the submission#<cell_number> compiler generated class that wraps the submission logic inside the cell.